core/src/main/java/org/apache/lucene/analysis/miscellaneous/DuplicateByteSequenceSpotter.java

```diff
@@ -0,0 +1,325 @@
+ * directly into the densely packed array using a byte value. The third level in
+ * the tree holds {@link LightweightTreeNode} nodes that have few children
+ * (typically much less than 256) and so use a dynamically-grown array to hold
+ * child objects of the type {@link ByteSequenceLeafNode}. These represent the
+ * final 3 bytes of a sequence as single int and hold a count of the number of
+ * times the entire sequence path has been visited.
+ * <p>
+ * The Trie grows indefinitely as more content is added and while theoretically
+ * it could be massive (a 6-depth tree could produce 256^6 nodes) non-random
+ * content e.g English text contains fewer variations.
+ * <p>
+ * In future we may look at using one of these strategies when memory is tight:
+ * <ol>
+ * <li>auto-pruning methods to remove less-visited parts of the tree
+ * <li>auto-reset to wipe the whole tree and restart when a memory threshold is
+ * reached
+ * <li>halting any growth of the tree
+ * </ol>
+ * 
+ * 
+ */
+public class DuplicateByteSequenceSpotter {
+    private static final int TREE_DEPTH = 6;
+    private final TreeNode root;
+    private boolean sequenceBufferFilled = false;
+    private final byte[] sequenceBuffer = new byte[TREE_DEPTH];
+    private final byte[] lastSequenceSearchBuffer = new byte[4];
+    private int nextFreePos = 0;
+
+    // ==Performance info
+    private final int[] nodesAllocatedByDepth;
+    private int nodesResizedByDepth;
+    // ==== RAM usage estimation settings ====
+    private long bytesAllocated;
+    // Root node object plus inner-class reference to containing "this"
+    // (profiler suggested this was a cost)
+    static final long TREE_NODE_OBJECT_SIZE = RamUsageEstimator.NUM_BYTES_OBJECT_HEADER + RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+    // A TreeNode specialization with an array ref (dynamically allocated and
+    // fixed-size)
+    static final long ROOT_TREE_NODE_OBJECT_SIZE = TREE_NODE_OBJECT_SIZE + RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+    // A KeyedTreeNode specialization with an array ref (dynamically allocated
+    // and grown)
+    static final long LIGHTWEIGHT_TREE_NODE_OBJECT_SIZE = TREE_NODE_OBJECT_SIZE + RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+    // A KeyedTreeNode specialization with a short-based hit count and a
+    // sequence of bytes encoded as an int
+    static final long LEAF_NODE_OBJECT_SIZE = TREE_NODE_OBJECT_SIZE + Short.BYTES + Integer.BYTES;
+
+    public DuplicateByteSequenceSpotter() {
+        this.nodesAllocatedByDepth = new int[4];
+        this.bytesAllocated = 0;
+        root = new RootTreeNode((byte) 1, null, 0);
+    }
+
+    /**
+     * Reset the sequence detection logic to avoid any continuation of the
+     * immediately previous bytes. A minimum of dupSequenceSize bytes need to be
+     * added before any new duplicate sequences will be reported.
+     */
+    public void startNewSequence() {
+        sequenceBufferFilled = false;
+        nextFreePos = 0;
```

Comment:
No, this is analogous to when you add an artificial postitionIncrement to tokens in Lucene to avoid false positives on phrase matches. It introduces a split/separator into a stream of text, typically when you end feeding it one doc and start with another. It doesn't influence the hit counts that have been accumulated so far. I actually found I needed it when I wrote the unit tests and found I was getting unintended results caused by "wrap arounds" from the end of one test doc to the start of the next.